src/storage/src/source/mysql/replication.rs

```diff
@@ -110,84 +118,113 @@ pub(crate) fn render<G: Scope<Timestamp = TransactionId>>(
+                    .filter(|t| t.timestamp() != &TransactionId::Absent)
+                    // Advance each partition's upper as far as the start transaction-id of the corresponding GTID UUID available in the binlog
+                    .map(|t| {
+                        let part = t.interval().singleton().expect("filtered non-singletons");
+                        let other = match binlog_start_gtids.get(part) {
+                            Some(binlog_start_gtid) => binlog_start_gtid.clone(),
+                            None => GtidPartition::new_singleton(part.clone(), TransactionId::Absent),
+                        };
+                        if t > &other {
+                            // TODO: Can we avoid this clone? Unfortunately there doesn't seem to be a way to take the element from an Antichain
+                            t.clone()
+                        } else {
+                            other
+                        }
+                    })
             );
 
-            let Some(resume_transaction_id) = resume_upper.into_option() else {
-                return Ok(());
-            };
-            data_cap_set.downgrade([&resume_transaction_id]);
-            upper_cap_set.downgrade([&resume_transaction_id]);
-            trace!(%id, "timely-{worker_id} replication reader started transaction_id={}", resume_transaction_id);
+            // Create capabilities for each active partition that we might see events for, and generate a frontier of all
+            // future partitions that may exist (the UUID gaps between the active partitions).
+            let mut active_partition_caps = BTreeMap::new();
+            let mut future_partitions = Antichain::new();
+            let mut prev = Uuid::nil();
+            for ts in resume_upper.iter().sorted_by(|a, b| a.interval().singleton().unwrap().cmp(b.interval().singleton().unwrap())) {
+                let part = ts.interval().singleton().unwrap();
+                assert!(prev < *part, "duplicate or unsorted partitions");
+
+                // Create a partition representing all the UUIDs in the gap between the previous one and this one
+                let gap = GtidPartition::new_range(prev, part.before(), TransactionId::Absent);
+                future_partitions.insert(gap);
+
+                // Set prev to the UUID + 1 so the next gap starts after this one
+                prev = part.after();
+
+                // Mint capabilities for this partition at the resume transaction-id
+                let part_ts = GtidPartition::new_singleton(part.clone(), ts.timestamp().clone());
+                trace!(%id, "timely-{worker_id} creating partition capability for {part_ts:?}");
+                let cap = GtidPartitionCapability::new(&part_ts, data_cap_set, upper_cap_set);
+                active_partition_caps.insert(part.clone(), cap);
+            }
+            future_partitions.insert(GtidPartition::new_range(prev, Uuid::maximum(), TransactionId::Absent));
```

Comment:
Assuming you follow the advice above about the definition of `resume_upper`, this becomes much more natural
```rust
// Convert the resume upper into the corresponding capabilities
let mut active_caps = HashMap::new();
let mut future_caps = Vec::new();
for ts in resume_upper {
    let caps = (data_cap_set.delayed(&ts), upper_cap_set.delayed(&ts);
    if ts.timestamp() == &TransactionId::Absent {
        future_caps.push(caps);
    } else {
        let part = ts.singleton().unwrap();
        active_caps.insert(part, caps);
    }
}
```